### PR TITLE
Increased width and height of the Campaign Difficulty selection dialog

### LINKED ISSUE #7511: Increase width and height of the Campaign Difficulty selection dialog
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

The campaign difficulty selection dialog has very limited space and should perhaps adjust its height dynamically to allow for languages that have longer texts. Additionally there should be more horizontal space since a couple of the supported languages have a bit longer difficulty level names than English:

image
image

### Describe the possible solution

Make the dialog wider and taller, preferrably adjusting its vertical height as needed.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by roman-yagodin**
This one resolves #7511.

## Changes:

- Splitted difficulty icons for separate rendering to allow further changes.
- Recalculated things based on difficulty icon size and its shadow size.
- Increased dialog width, slightly inreased top margin for description text and also dialog height.

## Screenshots:

English, good interface:

camp_difficulty_good_en_1

Bulgarian, evil interface:

camp_difficulty_evil_bg_1

Some difficulty options disabled (test):

camp_difficulty_good_bg_dis_3

camp_difficulty_evil_bg_dis_2

Dialog window width increased to 600 (test):

camp_difficulty_good_en_2

### CODE REVIEW

File: src/fheroes2/game/game_campaign.cpp [hunk 1]
```diff
@@ -907,14 +908,35 @@ namespace
...
+        const fheroes2::Sprite chessIcon = fheroes2::AGG::GetICN( ICN::NGHSBKG, 0 );
+
+        fheroes2::Copy( chessIcon, copyFromArea[0].x, copyFromArea[0].y, display, copyToOffset[0].x, copyToOffset[0].y, copyFromArea[0].width, copyFromArea[0].height );
+        fheroes2::Copy( chessIcon, copyFromArea[1].x, copyFromArea[1].y, display, copyToOffset[1].x, copyToOffset[1].y, copyFromArea[1].width, copyFromArea[1].height );
+        fheroes2::Copy( chessIcon, copyFromArea[2].x, copyFromArea[2].y, display, copyToOffset[2].x, copyToOffset[2].y, copyFromArea[2].width, copyFromArea[2].height );
 
         if ( isEvilInterface ) {
-            fheroes2::ApplyPalette( display, copyToOffset.x, copyToOffset.y, display, copyToOffset.x, copyToOffset.y, copyFromArea.width, copyFromArea.height,
-                                    PAL::GetPalette( PAL::PaletteType::GOOD_TO_EVIL_INTERFACE ) );
+            const std::vector<uint8_t> goodToEvilPalette = PAL::GetPalette( PAL::PaletteType::GOOD_TO_EVIL_INTERFACE );
```
Comment (ihhub) [inline src/fheroes2/game/game_campaign.cpp:932]: Please make this variable a const reference. Right no we are making a copy of the container.

File: src/fheroes2/game/game_campaign.cpp [hunk 2]
```diff
@@ -907,14 +908,35 @@ namespace
...
+
+            fheroes2::ApplyPalette( display, copyToOffset[0].x, copyToOffset[0].y, display, copyToOffset[0].x, copyToOffset[0].y, copyFromArea[0].width,
+                                    copyFromArea[0].height, goodToEvilPalette );
+            fheroes2::ApplyPalette( display, copyToOffset[1].x, copyToOffset[1].y, display, copyToOffset[1].x, copyToOffset[1].y, copyFromArea[1].width,
+                                    copyFromArea[1].height, goodToEvilPalette );
+            fheroes2::ApplyPalette( display, copyToOffset[2].x, copyToOffset[2].y, display, copyToOffset[2].x, copyToOffset[2].y, copyFromArea[2].width,
+                                    copyFromArea[2].height, goodToEvilPalette );
```
Comment (ihhub) [inline src/fheroes2/game/game_campaign.cpp:939]: We can reduce code duplication and avoid hardcoded values by doing this:
```cpp
for ( size_t i = 0; i < copyToOffset.size(); ++i ) {
    fheroes2::ApplyPalette( display, copyToOffset[i].x, copyToOffset[i].y, display, copyToOffset[i].x, copyToOffset[i].y, copyFromArea[i].width,
        copyFromArea[i].height, goodToEvilPalette );
}
```

The same actually applies to lines 927-929.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @roman-yagodin , I left 2 small optimization related comments. Would you mind to take a look once you have time?

**Comment by roman-yagodin** [comment]
Hi, @ihhub! Sure.

**Review by ihhub — APPROVED** [review]

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@roman-yagodin , thank you so much for you contribution to the project!

**Comment by ihhub** [comment]
@roman-yagodin , no worries. I did the changes because we are getting close to the release date and I would like to have more changes to be merged.

**Pull request merged**